src/connector/src/parser/scalar_adapter.rs

```diff
@@ -157,80 +176,142 @@ impl ScalarAdapter<'_> {
             }
             (ScalarAdapter::Enum(EnumString(s)), &DataType::Varchar) => Some(ScalarImpl::from(s)),
-            _ => {
-                tracing::error!(
-                    adapter = self.name(),
-                    rw_type = ty.pg_name(),
-                    "failed to convert from ScalarAdapter: invalid conversion"
-                );
-                None
-            }
-        }
-    }
-
-    pub fn build_scalar_in_list(
-        vec: Vec<Option<ScalarAdapter<'_>>>,
-        ty: DataType,
-        mut builder: risingwave_common::array::ArrayBuilderImpl,
-    ) -> Option<risingwave_common::array::ArrayBuilderImpl> {
-        for val in vec {
-            let scalar = match (val, &ty) {
-                (Some(ScalarAdapter::Numeric(numeric)), &DataType::Varchar) => {
-                    if pg_numeric_is_special(&numeric) {
-                        return None;
-                    } else {
-                        Some(ScalarImpl::from(pg_numeric_to_string(&numeric)))
-                    }
-                }
-                (Some(ScalarAdapter::Numeric(numeric)), &DataType::Int256) => match numeric {
-                    PgNumeric::Normalized(big_decimal) => {
-                        match Int256::from_str(big_decimal.to_string().as_str()) {
-                            Ok(num) => Some(ScalarImpl::from(num)),
-                            Err(err) => {
-                                tracing::error!(error = %err.as_report(), "parse pg-numeric as rw_int256 failed");
+            (ScalarAdapter::NumericList(vec), &DataType::List(dtype)) => {
+                let mut builder = dtype.create_array_builder(0);
+                for val in vec {
+                    let scalar = match (val, &dtype) {
+                        (Some(numeric), box DataType::Varchar) => {
+                            if pg_numeric_is_special(&numeric) {
                                 return None;
+                            } else {
+                                Some(ScalarImpl::from(pg_numeric_to_string(&numeric)))
                             }
                         }
-                    }
-                    _ => return None,
-                },
-                (Some(ScalarAdapter::Numeric(numeric)), &DataType::Decimal) => match numeric {
-                    PgNumeric::Normalized(big_decimal) => {
-                        match Decimal::from_str(big_decimal.to_string().as_str()) {
-                            Ok(num) => Some(ScalarImpl::from(num)),
-                            Err(err) => {
-                                tracing::error!(error = %err.as_report(), "parse pg-numeric as rw-numeric failed (likely out-of-range");
-                                return None;
+                        (Some(numeric), box DataType::Int256) => match numeric {
+                            PgNumeric::Normalized(big_decimal) => {
+                                match Int256::from_str(big_decimal.to_string().as_str()) {
```

Comment:
This part is duplicate of `pg_numeric_to_rw_int256`, I think you can also call `into_scalar(dtype)` recursively to reduce redundant code.

Reply:
updated.